### PR TITLE
Create new rule/why-upgrade-to-latest-dotnet

### PR DESCRIPTION
**Pull request by YazhiChen**
As per my conversation with @adamcogan on Oct4 (last Wed), I will need to create a rule discussing 'Why do we want to move to .NET 8? Why is .NET Framework evil?'. A rule similar to this rule: [Do you know why to upgrade from AngularJS to Angular?](https://www.ssw.com.au/rules/why-upgrade-to-latest-angular/)

I have discussed the pain points of the legacy .NET Framework and the benefits for migrating to the latest .NET in this rule.

### CODE REVIEW

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:25]: Change colons to dashes, it's easier to notice the split and aligns with the prefixes rules 

Take a look at https://www.ssw.com.au/rules/prefixes/

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:35]: It looks like the intent here is headers, we should use h4's maybe, instead of bullets.

Also markdown like empty lines as separators so add in an extra 1 before and after the para to make sure it does the breaks as you'd expect

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:25]:
```suggestion
* Compatibility: .NET Framework applications run on Windows, limiting cross-platform deployment.
```

Any .NET Framework not running on windows is likely mono (i.e.: Not .NET Framework) or some other flavor of windows. Unless you have a resource that suggests otherwise? Maybe I'm wrong?

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:26]: Phrase this more positively?

```suggestion
* Performance: Each new version of .NET brings improved performance and resource utilization, allowing the same .NET code to perform better on the same hardware.
```

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:27]: This line doesn't feel true

> Beginning with version 4.5.2 and later, .NET Framework is defined as a component of the Windows operating system (OS). Components receive the same support as their parent products, therefore, .NET Framework 4.5.2 and later follows the lifecycle policy of the underlying Windows OS on which it is installed.

https://learn.microsoft.com/en-us/lifecycle/faq/dotnet-framework

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:28]: Phrase positively?

```suggestion
* Modern development features: Newer versions of .NET give us more opportunities to design products and services using many newer capabilities that are not backported to .NET Framework. This allows us to build, test and deploy faster, and more securely than ever before, keeping us competitive in today's tech landscape.
```

You might want to double-check the line to make sure I've not gone too far to the other side 😅

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:46]: End off with a link to .NET Framework support page

https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework

and .NET support page

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:38]: > This widens your reach and allows you to target a broader audience.

This sentence doesn't sit well with me because we're generally not building thick clients anymore. So our web apps allow us to target all customers already regardless of operating system.

It does, however, give us more flexibility in deployment, opening up the world of containerization, as well as using the same libraries in Web, WebAssembly (.NET Blazor), and native mobile apps using .NET MAUI

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:40]: Refer to 1 of my other points. This line is inaccurate as .NET Framework is supported full under windows, so if there were any security issues it would be patched as part of windows

https://learn.microsoft.com/en-us/lifecycle/products/microsoft-net-framework

https://learn.microsoft.com/en-us/lifecycle/faq/dotnet-framework#what-is-the-microsoft-lifecycle-policy-for--net-framework-

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:42]: C# 11 is out, and 12 comes out next month.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version#defaults

Also, 90% of C# features can be used in .NET Framework as they are syntactical sugar

To try this out, add this to any .NET Framework project, I did this just today to make use of the cleaner using statements

```xml
<PropertyGroup>
   <LangVersion>11.0</LangVersion>
</PropertyGroup>
```

Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:42]: I don't think it's relevant saying .NET lets you use blazor, so you should upgrade... The support path unless you're adding blazor server to an existing app (I don't think that's recommended) is creating a new app regardless, so your old app can be .NET Framework while a new and shiny blazor app exists in the product

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:44]: I'd remove on going support as per some other comments (because it's there for .NET Framework)

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:25]: You are right. Double checked the Microsoft Doc.

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:26]: This part is talking about the pain point of .NET framework, why we do not want to use it anymore.
And in the later part - ### Enter the Modern .NET(## Key Reasons to Upgrade), I am talking about the .NET, why do we want to use .NET to make the comparison.

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:27]: Can I rephrase this to be "Some older versions of .NET Framework reached a point where it was no longer actively developed and supported, leaving applications vulnerable to security risks"? As said in the article, support for some older versions are ended in 2016 etc...

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:28]: This is the same, in this section, I want to focus on the pain point of the .NET Framework and why we do not want to stay in .NET Framework anymore.

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [reply rules/why-upgrade-to-latest-dotnet/rule.md:26]: Sure, you can say that .NET Framework is slow by saying newer .NET is fast 🙂. 

The way it's phrased at the moment, .NET 4.8.1 is faster and less resource intensive... because after that is .NET Core and then .NET, `older versions of .NET Framework` excludes the new versions of .NET Framework.

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [reply rules/why-upgrade-to-latest-dotnet/rule.md:27]: If you say that, why would I upgrade to .NET 8 instead of just upgrading to .NET Framework 4.8.1? It's still supported and my business case clearly doesn't need other features otherwise I would have moved already🙂

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [reply rules/why-upgrade-to-latest-dotnet/rule.md:28]: It's just a suggestion 🙂. You can illustrate pain points through a positive for the alternative. 

Saying things like may not support modern development features makes it sound like you don't know if it does or not, I'd then make that be more assertive, as in, it doesn't support modern development features and patterns. 
Although the sentence is not completely accurate, as it depends on the business use case. 

This is where showing the alternative as the pain point comes across better

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Added a summary section

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated.

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md]: Updated

**Review by christoment — APPROVED** [review]
As @YazhiChen  has implemented the changes discussed with me, this LGTM. But please wait for approval from @GordonBeeming  😉

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:37]: @YazhiChen these should be h4's, h3 doesn't style right on the pages

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (GordonBeeming) [inline rules/why-upgrade-to-latest-dotnet/rule.md:26]: @YazhiChen we could maybe bold the prefixes so they stand out better on the page. At the moment it looks like a wall of text and I think this might help

Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:26]: Hi @GordonBeeming, this is fixed in the PR #6959

File: rules/why-upgrade-to-latest-dotnet/rule.md
Comment (YazhiChen) [reply rules/why-upgrade-to-latest-dotnet/rule.md:37]: Hi @GordonBeeming, this is fixed in the PR #6959

**Pull request merged**
